### ISSUE TITLE
Inconsistent Cluster Assignments in KMeans

### ISSUE THREAD

**Issue by HendrikHuel** [wannesm/dtaidistance#139]
I try to extend the KMeans clustering algorithm with an predict function. While doing so I encountered a strange problem while comparing the cluster labels from the fitted model to my predictions. The predicted labels are calculated quite similar to the procedure in KMeans.fit().

Here is my code.

```
import numpy as np
import multiprocessing as mp

import tslearn.datasets.ucr_uea as tsld
from dtaidistance import clustering as dtaic

# Time series data from tslearn for benchmarking
series = tsld.UCR_UEA_datasets().load_dataset("TwoPatterns")[0][:, :, 0]

# Init the KMeans
model = dtaic.KMeans(
    k=4,
    max_it=2,
    max_dba_it=100,
    thr=1e-6,
    dists_options={"window": 2},
    initialize_sample_size=5,
)

# Fit the KMeans
cluster_idx, performed_it = model.fit(series, use_c=True, use_parallel=True)

# Gather labels
labels_from_model_fit = np.ones(len(series)) * np.nan
for c in cluster_idx.keys():
    cluster_center = model.means[c]
    labels_from_model_fit[list(cluster_idx[c])] = c


def predict(series, model):
    """
    Find closest cluster. Similar to the distance calculation during KMeans.fit().
    """
    fn = dtaic.kmeans._distance_c_with_params
    with mp.Pool() as p:
        clusters_distances = p.map(
            fn,
            [
                (series[idx], model.means, model.dists_options)
                for idx in range(len(series))
            ],
        )

    return np.array(clusters_distances)[:, 0]

# Assign each time-series to a cluster
labels_prediction_like = predict(series, model)

# Count differences
print(np.sum((labels_from_model_fit - labels_prediction_like) != 0))
```

The printed sum will be lager than zero. So there is a difference between predicted labels and model labels.

Is there an obvious mistake?

**Comment by HendrikHuel** [comment]
After studying the code of `KMeans.fit()` in `kmeans.py` I might have found the bug or, rather, the reason for the inconsistency between labels and cluster centers.

1. The algorithm first calculates the distance between the old or intialized mediods and assigns each series to a cluster.
2. The algorithm then calculates the new cluster means.
3. Since the new means can differ from the means used in step 1. the implizit cluster labels can change

To fix this one has simply to swap step 1. and 2. and run an initial calculation of `cluster_distances` with the intial kmoids before the loop. If there is some interest in this fix, I could start a pull request.

**Comment by wannesm** [comment]
Thanks for the effort @HendrikHuel , good analysis. That is indeed a missing assignment step. It is fixed it in 1cb821a .
